latency.go

```diff
@@ -0,0 +1,93 @@
+package l2
+
+import (
+	"log"
+	"time"
+)
+
+// This is a wrapper that can take a device and limit the bandwidth that can
+// go through it.
+type deviceWithLatency struct {
+	dev FrameReadWriteCloser
+
+	// Maximum number of outgoing bytes per second.
+	send_bandwidth int
+	// Maximum number of incoming bytes per second.
+	receive_bandwidth int
+	// How much data we wrote the last time we called WriteFrame().
+	last_write_size int
+	// How much data we read the last time we called ReadFrame().
+	last_read_size int
+}
+
+// Basically initializes latency and then delegates to the tapDevice version of
+// NewTapDevice.
+// Args:
+//	mac: MAC address of the device.
+//  dev: A name for the device.
+//	send_bandwidth: Maximum number of outgoing bytes per second.
+//  receive_bandwidth: Maximum number of incoming bytes per second.
+// Returns:
+//	The new tap device, error.
+func NewTapDeviceWithLatency(mac, dev string, send_bandwidth,
+	receive_bandwidth int) (FrameReadWriteCloser, error) {
+	tap, err := NewTapDevice(mac, dev)
+	if err != nil {
+		return nil, err
+	}
+
+	var wrapped_tap deviceWithLatency
```

Comment:
I'm not sure if this will zero initialize, but either way given that you want to fill in some fields

wrapped_tap := deviceWithLatency{
dev: tap
send_bandwidth: send_bandwidth
receive_bandwidth: receive_bandwidth
}